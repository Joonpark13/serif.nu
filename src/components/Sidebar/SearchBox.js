import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core/styles';
import { debounce } from 'debounce';

export const styles = {
  container: {
    margin: 15,
  },
};

class SearchBox extends Component {
  constructor(props) {
    super(props);

    this.state = {
      searchText: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleChangeHelper = debounce(this.handleChangeHelper.bind(this), 300);
  }

  handleChange(value) {
    const { clearSearchResults } = this.props;

    this.setState({ searchText: value });

    if (value.length > 2) {
      this.handleChangeHelper(value);
    } else {
      clearSearchResults();
    }
  }

  // Added because debouncing the handleChange function
  // delays the text showing in the textbox (setState)
  handleChangeHelper(value) {
    const { handleSearchInput } = this.props;

    handleSearchInput(value);
  }

  render() {
    const { classes } = this.props;
    const { searchText } = this.state;

    return (
      <div className={classes.container}>
        <TextField
          value={searchText}
          onChange={event => this.handleChange(event.target.value)}
          label="Search for classes"
          type="search"
          fullWidth
        />
      </div>
    );
  }
}

export { SearchBox as UnstyledSearchBox };

export default withStyles(styles)(SearchBox);

SearchBox.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  handleSearchInput: PropTypes.func.isRequired,
  clearSearchResults: PropTypes.func.isRequired,
};
