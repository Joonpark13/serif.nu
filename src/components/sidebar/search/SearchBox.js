import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { debounce } from 'debounce';
import { withStyles } from '@material-ui/styles';
import { TextField } from '@material-ui/core';
import { currentSearchInputSelector } from 'selectors';
import { fetchSearchResultsRequest, clearSearchResults, updateSearchInput } from 'actions';
import useSelector from 'util/use-selector';

export const styles = {
  container: {
    margin: 15,
  },
};

function SearchBox({ classes }) {
  const currentSearchInput = useSelector(currentSearchInputSelector);
  const dispatch = useDispatch();

  function handleChange(value) {
    dispatch(updateSearchInput(value));

    if (value.length > 2) {
      debounce(() => dispatch(fetchSearchResultsRequest(value)), 300)();
    } else {
      dispatch(clearSearchResults());
    }
  }

  return (
    <div className={classes.container}>
      <TextField
        value={currentSearchInput}
        onChange={event => handleChange(event.target.value)}
        label="Search for classes"
        type="search"
        fullWidth
      />
    </div>
  );
}

export { SearchBox as UnstyledSearchBox };

export default withStyles(styles)(SearchBox);

SearchBox.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};
