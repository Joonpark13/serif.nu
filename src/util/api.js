/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
