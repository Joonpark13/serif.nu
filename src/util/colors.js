export const northwesternPurple = '#4E2A84';
export const northwesternBrightOrange = '#FFC520';
export const northwesternRichBlack10 = '#D8D6D6';
