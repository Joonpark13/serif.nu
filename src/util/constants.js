/* eslint-disable import/prefer-default-export */
export const CURRENT_TERM = '4730';
