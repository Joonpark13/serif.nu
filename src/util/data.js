import termsData from 'data/terms.json';

/* eslint-disable import/prefer-default-export */
/* istanbul ignore next */
export const currentTerm = termsData.reduce((highestTerm, term) => {
  if (term.id > highestTerm.id) {
    return term;
  }
  return highestTerm;
});
/* eslint-enable import/prefer-default-export */
