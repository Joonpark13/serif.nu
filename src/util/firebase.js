import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

firebase.initializeApp({
  apiKey: 'AIzaSyDluAG6bnqleqJ3yjPvGdlb3RyZP3U5JIQ',
  authDomain: 'sans-serif-northwestern.firebaseapp.com',
  databaseURL: 'https://sans-serif-northwestern.firebaseio.com',
  projectId: 'sans-serif-northwestern',
  storageBucket: 'sans-serif-northwestern.appspot.com',
  messagingSenderId: '1078100635889',
});

export const db = firebase.firestore();
export const auth = firebase.auth();
