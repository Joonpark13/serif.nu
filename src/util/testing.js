/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import * as useSelector from './use-selector';

jest.mock('./use-selector');

// Helper function that returns a function which creates enzyme wrappers with default props
export function wrapperCreator(Component, defaultProps = {}) {
  let classes;
  return props => shallow(<Component classes={classes} {...defaultProps} {...props} />);
}

/* istanbul ignore next */
export function mockUseSelector(...args) {
  useSelector.default.mockReset();
  args.forEach(arg => useSelector.default.mockReturnValueOnce(arg));

  return useSelector.default;
}

/* istanbul ignore next */
export function mockUseDispatch() {
  const dispatchSpy = jest.fn();
  reactRedux.useDispatch.mockReturnValue(dispatchSpy);
  return dispatchSpy;
}
