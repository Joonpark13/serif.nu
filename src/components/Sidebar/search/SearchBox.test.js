import React from 'react';
import { shallow } from 'enzyme';
import { mockStyles } from 'util/testing';
import { UnstyledSearchBox, styles } from './SearchBox';

describe('SearchBox', () => {
  function makeEventObject(value) {
    // This object gets passed to the onChange function of TextField
    return { target: { value } };
  }
  const testSearchInput = 'EECS';

  it('renders correctly', () => {
    const classes = mockStyles(styles);
    const wrapper = shallow(<UnstyledSearchBox
      classes={classes}
      handleSearchInput={() => {}}
      clearSearchResults={() => {}}
      updateSearchInput={() => {}}
      currentSearchInput={testSearchInput}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders text typed into the search box', () => {
    const classes = mockStyles(styles);
    const wrapper = shallow(<UnstyledSearchBox
      classes={classes}
      handleSearchInput={() => {}}
      clearSearchResults={() => {}}
      updateSearchInput={() => {}}
      currentSearchInput={testSearchInput}
    />);

    const eventObject = makeEventObject(testSearchInput);
    wrapper.find('TextField').simulate('change', eventObject);

    expect(wrapper).toMatchSnapshot();
  });

  it('doesn\'t call api unless it has been 300ms since last keystroke', (completed) => {
    const handleSearchInputMock = jest.fn();

    const classes = mockStyles(styles);
    const wrapper = shallow(<UnstyledSearchBox
      classes={classes}
      handleSearchInput={handleSearchInputMock}
      clearSearchResults={() => {}}
      updateSearchInput={() => {}}
      currentSearchInput={testSearchInput}
    />);

    expect(handleSearchInputMock).not.toBeCalled();

    const eventObject = makeEventObject(testSearchInput);
    wrapper.find('TextField').simulate('change', eventObject);

    expect(handleSearchInputMock).not.toBeCalled();

    // Async test, would end after this statement but then the expect wouldn't
    // contribute to test results
    setTimeout(((jestCompleted) => {
      expect(handleSearchInputMock).toHaveBeenCalledWith(testSearchInput);
      jestCompleted();
    }).bind(null, completed), 300);
  });

  it('doesn\'t call api through handler if search string <= 2 chars long', () => {
    const handleSearchInputMock = jest.fn();
    const classes = mockStyles(styles);
    const wrapper = shallow(<UnstyledSearchBox
      classes={classes}
      handleSearchInput={handleSearchInputMock}
      clearSearchResults={() => {}}
      updateSearchInput={() => {}}
      currentSearchInput={testSearchInput}
    />);

    const shortTestSearchInput = 'EE';
    const eventObject = makeEventObject(shortTestSearchInput);
    wrapper.find('TextField').simulate('change', eventObject);
    expect(wrapper).toMatchSnapshot();
  });
});
