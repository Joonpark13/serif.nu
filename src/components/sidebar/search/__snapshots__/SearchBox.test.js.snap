// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders correctly 1`] = `
<div
  className="makeStyles-container-1"
>
  <ForwardRef(WithStyles)
    fullWidth={true}
    label="Search for classes"
    onChange={[Function]}
    type="search"
    value="EECS"
  />
</div>
`;

exports[`SearchBox renders text typed into the search box 1`] = `
<div
  className="makeStyles-container-1"
>
  <ForwardRef(WithStyles)
    fullWidth={true}
    label="Search for classes"
    onChange={[Function]}
    type="search"
    value="EECS"
  />
</div>
`;
